migrations/20220818164120_create_anchor_and_request.js

```diff
@@ -0,0 +1,102 @@
+const REQUEST_TABLE_NAME = 'request'
+const TEMP_TABLE_NAME = 'temp'
+const ANCHOR_TABLE_NAME = 'anchor'
+
+const sortByStringKeys = (obj) =>
+  Object.fromEntries(Object.entries(obj).sort(([a], [b]) => a.localeCompare(b)))
+
+const checkObjEquality = (expected, received) =>
+  JSON.stringify(sortByStringKeys(expected)) === JSON.stringify(sortByStringKeys(received))
+
+const verifyTableStructure = async (knex, tableWithValidStructure, tableToCheck) => {
+  const [expectedStructure, existingStructure] = await Promise.all([
+    knex(tableWithValidStructure).columnInfo(),
+    knex(tableToCheck).columnInfo(),
+  ])
+  const structuresAreEqual = checkObjEquality(expectedStructure, existingStructure)
+  if (!structuresAreEqual) {
+    throw Error(`Existing ${tableToCheck} table does not have the expected structure`)
+  }
+}
+
+const createRequestTable = async (knex, tableName) => {
+  await knex.schema.createTable(tableName, (table) => {
+    table.uuid('id').notNullable().unique().primary().defaultTo(knex.raw('uuid_generate_v4()'))
+    table.integer('status').notNullable()
+    table.boolean('pinned').notNullable().defaultTo(false)
+    table.timestamp('created_at', { useTz: false }).notNullable().defaultTo(knex.raw('now()'))
+    table.timestamp('updated_at', { useTz: false }).notNullable().defaultTo(knex.raw('now()'))
+  })
+
+  // knex defaults to varchar(255).
+  // This is not best practise as noted here: https://wiki.postgresql.org/wiki/Don%27t_Do_This#Don.27t_use_varchar.28n.29_by_default
+  // knex has an open PR to fix this but as a workaround we use raw queries
+  await knex.schema
+    .raw(`alter table "${tableName}" add column "cid" varchar NOT NULL UNIQUE`)
+    .raw(`alter table "${tableName}" add column "doc_id" varchar NOT NULL`)
+    .raw(`alter table "${tableName}" add column "message" varchar`)
+}
+
+const createRequestTableIfNotExists = async (knex) => {
+  const tableExists = await knex.schema.hasTable(REQUEST_TABLE_NAME)
+
+  if (!tableExists) {
+    await createRequestTable(knex, REQUEST_TABLE_NAME)
+  } else {
+    await createRequestTable(knex, TEMP_TABLE_NAME)
+    await verifyTableStructure(knex, TEMP_TABLE_NAME, REQUEST_TABLE_NAME)
+    await knex.schema.dropTable(TEMP_TABLE_NAME)
+  }
+
+  // explicitly rename the doc_id column to stream_id
```

Comment:
won't this be a problem if the table already exists and has already renamed its column to stream_id?  Wouldn't it actually fail `verifyTableStructure` above?

Reply:
I don't know how knex migrations really work, but it seems like they should be able to have some conditional logic to only run if they haven't already run.  Can we pull out the changes to the table structures over what was there previously into their own migration - so first we create the tables as they've always been, then we do a second migration to update them into the new schemas, and make sure it all works when run against all 3 possible prior states (clean with no tables, has existing tables in the old format, has existing tables in the new format)

Reply:
yup good call. This will make it a lot cleaner